x/bank/keeper_test.go

```diff
@@ -0,0 +1,217 @@
+	coinKeeper := NewCoinKeeper(accountMapper)
+	sendKeeper := NewSendKeeper(accountMapper)
+
+	addr := sdk.Address([]byte("addr1"))
+	addr2 := sdk.Address([]byte("addr2"))
+	addr3 := sdk.Address([]byte("addr3"))
+	acc := accountMapper.NewAccountWithAddress(ctx, addr)
+
+	// Test GetCoins/SetCoins
+	accountMapper.SetAccount(ctx, acc)
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{}))
+
+	coinKeeper.SetCoins(ctx, addr, sdk.Coins{{"foocoin", 10}})
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{{"foocoin", 10}}))
+
+	// Test HasCoins
+	assert.True(t, sendKeeper.HasCoins(ctx, addr, sdk.Coins{{"foocoin", 10}}))
+	assert.True(t, sendKeeper.HasCoins(ctx, addr, sdk.Coins{{"foocoin", 5}}))
+	assert.False(t, sendKeeper.HasCoins(ctx, addr, sdk.Coins{{"foocoin", 15}}))
+	assert.False(t, sendKeeper.HasCoins(ctx, addr, sdk.Coins{{"barcoin", 5}}))
+
+	coinKeeper.SetCoins(ctx, addr, sdk.Coins{{"foocoin", 15}})
+
+	// Test SendCoins
+	sendKeeper.SendCoins(ctx, addr, addr2, sdk.Coins{{"foocoin", 5}})
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{{"foocoin", 10}}))
+	assert.True(t, sendKeeper.GetCoins(ctx, addr2).IsEqual(sdk.Coins{{"foocoin", 5}}))
+
+	err2 := sendKeeper.SendCoins(ctx, addr, addr2, sdk.Coins{{"foocoin", 50}})
+	assert.Implements(t, (*sdk.Error)(nil), err2)
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{{"foocoin", 10}}))
+	assert.True(t, sendKeeper.GetCoins(ctx, addr2).IsEqual(sdk.Coins{{"foocoin", 5}}))
+
+	coinKeeper.AddCoins(ctx, addr, sdk.Coins{{"barcoin", 30}})
+	sendKeeper.SendCoins(ctx, addr, addr2, sdk.Coins{{"barcoin", 10}, {"foocoin", 5}})
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{{"barcoin", 20}, {"foocoin", 5}}))
+	assert.True(t, sendKeeper.GetCoins(ctx, addr2).IsEqual(sdk.Coins{{"barcoin", 10}, {"foocoin", 10}}))
+
+	// Test InputOutputCoins
+	input1 := NewInput(addr2, sdk.Coins{{"foocoin", 2}})
+	output1 := NewOutput(addr, sdk.Coins{{"foocoin", 2}})
+	sendKeeper.InputOutputCoins(ctx, []Input{input1}, []Output{output1})
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{{"barcoin", 20}, {"foocoin", 7}}))
+	assert.True(t, sendKeeper.GetCoins(ctx, addr2).IsEqual(sdk.Coins{{"barcoin", 10}, {"foocoin", 8}}))
+
+	inputs := []Input{
+		NewInput(addr, sdk.Coins{{"foocoin", 3}}),
+		NewInput(addr2, sdk.Coins{{"barcoin", 3}, {"foocoin", 2}}),
+	}
+
+	outputs := []Output{
+		NewOutput(addr, sdk.Coins{{"barcoin", 1}}),
+		NewOutput(addr3, sdk.Coins{{"barcoin", 2}, {"foocoin", 5}}),
+	}
+	sendKeeper.InputOutputCoins(ctx, inputs, outputs)
+	assert.True(t, sendKeeper.GetCoins(ctx, addr).IsEqual(sdk.Coins{{"barcoin", 21}, {"foocoin", 4}}))
+	assert.True(t, sendKeeper.GetCoins(ctx, addr2).IsEqual(sdk.Coins{{"barcoin", 7}, {"foocoin", 6}}))
+	assert.True(t, sendKeeper.GetCoins(ctx, addr3).IsEqual(sdk.Coins{{"barcoin", 2}, {"foocoin", 5}}))
+
+}
+
+func TestViewKeeper(t *testing.T) {
```

Comment:
Add tests to make sure `ViewKeeper` can't be cast to `SendKeeper`.

Reply:
They are currently both structs, so can't cast them to each other already.  Won't compile.